src/main/java/com/iota/iri/service/API.java

```diff
@@ -762,37 +765,44 @@ public void broadcastTransactionStatement(final List<String> trytes2) {
         }
     }
 
-    private AbstractResponse getBalancesStatement(final List<String> addrss, final int threshold) throws Exception {
+    private AbstractResponse getBalancesStatement(final List<String> addrss, final List<String> txs, final int threshold) throws Exception {
 
         if (threshold <= 0 || threshold > 100) {
             return ErrorResponse.create("Illegal 'threshold'");
         }
 
         final List<Hash> addresses = addrss.stream().map(address -> (new Hash(address)))
                 .collect(Collectors.toCollection(LinkedList::new));
+        final List<Hash> hashes;
+        if (txs == null || txs.size() == 0) {
+            hashes = Collections.singletonList(instance.milestone.latestSolidSubtangleMilestone);
+        } else {
+            hashes = txs.stream().map(address -> (new Hash(address)))
+                    .collect(Collectors.toCollection(LinkedList::new));
+        }
 
         final Map<Hash, Long> balances = new HashMap<>();
         final int index;
+        Snapshot referenceSnapshot;
         synchronized (instance.milestone.latestSnapshot.snapshotSyncObject) {
-            index = instance.milestone.latestSnapshot.index();
-            for (final Hash address : addresses) {
-                balances.put(address,
-                        instance.milestone.latestSnapshot.getState().containsKey(address) ?
-                                instance.milestone.latestSnapshot.getState().get(address) : Long.valueOf(0));
+            referenceSnapshot = new Snapshot(instance.milestone.latestSnapshot);
+        }
+        index = referenceSnapshot.index();
+        for(Hash hash: hashes) {
+            if (!instance.ledgerValidator.isTipConsistent(referenceSnapshot, hash)) {
+                return ErrorResponse.create("Hashes are not consistent");
```

Comment:
non-existing/yet-to-be-seen hashes also return `Hashes are not consistent` - this will be used for newly transmitted transactions, so getting this error message might be confusing.

an explicit check that the hashes are present would be best.